### PR TITLE
 buildRustPackage: support cross-compilation to x86_64-pc-mingw32

### PR DESCRIPTION
**Pull request by lukateras**
#### Motivation for this change

Allows to build Rust packages on Windows. Intended usage scenario is:

```nix
let
  nixpkgs-mozilla = import (fetchTarball {
    url = "https://github.com/mozilla/nixpkgs-mozilla/archive/ac8e9d7bbda8fb5e45cae20c5b7e44c52da3ac0c.tar.gz";
    sha256 = "1irlkqc0jdkxdfznq7r52ycnf0kcvvrz416qc7346xhmilrx2gy6";
  });
in

with import <nixpkgs> { overlays = [ nixpkgs-mozilla ]; };

let
  rustChannel = rustChannelOfTargets "nightly" "2019-07-14" [ "x86_64-pc-windows-gnu" ];
  rustPlatform = pkgsCross.mingwW64.makeRustPlatform {
    cargo = rustChannel;
    rustc = rustChannel;
  };
in

rustPlatform.buildRustPackage {
  name = "example";
  src = lib.cleanSource ./.;

  preConfigure = ''
    export HOME=$(mktemp -d)
  '';

  cargoSha256 = "0000000000000000000000000000000000000000000000000000";
}
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Ericson2314 @matthewbauer

### CODE REVIEW

**Comment by yegortimoshenko** [comment]
(cc @Mic92)

File: pkgs/build-support/rust/default.nix [hunk 1]
```diff
@@ -41,18 +41,25 @@ let
...
+
+stdenv.mkDerivation (args // {
   inherit cargoDeps;
 
   patchRegistryDeps = ./patch-registry-deps;
 
-  nativeBuildInputs = [ cargo rustc git cacert ] ++ nativeBuildInputs;
-  inherit buildInputs;
+  nativeBuildInputs = nativeBuildInputs ++ [ cacert git cargo rustc ];
+  buildInputs = buildInputs ++ stdenv.lib.optional (nixConfig == "x86_64-pc-mingw32") windows.pthreads;
```
Comment (Ericson2314) [inline pkgs/build-support/rust/default.nix]: isWindows or isMinGW?

File: pkgs/os-specific/windows/pthread-w32/default.nix [hunk 2]
```diff
@@ -1,61 +1,29 @@
...
+    runHook postInstall
+  '';
 
-    license = stdenv.lib.licenses.lgpl21Plus;
+  meta = with stdenv.lib; {
+    description = "POSIX threads library for Windows";
+    homepage = "https://sourceware.org/pthreads-win32";
+    license = licenses.lgpl21Plus;
+    maintainers = with maintainers; [ yegortimoshenko ];
+    platforms = platforms.windows;
```
Comment (Ericson2314) [inline pkgs/os-specific/windows/pthread-w32/default.nix]: CC @angerman was this all broken?

Comment (yegortimoshenko) [reply pkgs/os-specific/windows/pthread-w32/default.nix]: It probably was, `nix-build '<nixpkgs>' -A pkgsCross.mingwW64.windows.pthreads --no-out-link` fails on `No rule to make target 'w32api.h.in', needed by 'w32api.h'` as of recent master.

File: pkgs/build-support/rust/default.nix (hunk 1 above)
Comment (yegortimoshenko) [reply pkgs/build-support/rust/default.nix]: Changed to `stdenv.hostPlatform.isMinGW`, thanks! `pthreads` is unlikely to be a dependency in MSVC builds.

**Comment by yegortimoshenko** [comment]
@Ericson2314 <3

**Pull request merged**